### PR TITLE
Merge/continous block improvement

### LINKED ISSUE #3592: implement dynamic block building
Currently, we are able to trigger the building of a block on top of a known head, however if we want to create another block a few seconds later (to include potentially more profitable txs) we would have to recreate the block from scratch.

Geth includes a mechanism for continuous block improvement, where the currently mined block is continuously improved with better incoming txs and tx ordering optimizations. We should implement something similar.

This will also be very critical for us after the merge because of the way consensus clients request blocks, as well as because then we will be a mining client so we should aim to maximize the profits of validators running Nethermind.

### PR DESCRIPTION
**Pull request by LukaszRozmej**
Resolves #3592

This is based on 14.1 version. Merge ready if someone wants to use it.

## Changes:
- Adds the ability to constantly improve the block between ForkChoiceUpdated and GetPayload calls, this allows to create better and more profitable blocks if there is a larger gap between them. Between FCU and GetPayload can be realistically 7-10s, depending on CL, networking, and processing time. Currently picked delay between improvements to 3s, which could produce 3-4 blocks per slot with those timings.
- This is carefully designed to: 
1. Not leak block improvements - they are disposed, there is only one improvement per payload id. The underlying task is canceled on dispose.
2. On GetPayload if we already constructed non-empty block to return immediately 
3. Not to run further improvements when previous one was disposed or block slot has passed.
- Moved all payload creation tests to separate file

## Types of changes

What types of changes does your code introduce?
_Put an `x` in the boxes that apply_

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation Update
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] Other (please describe): 

## Testing
**Requires testing**

- [x] Yes
- [ ] No

**In case you checked yes, did you write tests??**

- [x] Yes
- [ ] No

**Comments about testing , should you have some**
Should be tested on shadow-fork. Would be good to grab memory dump to ensure there is no leak.

### CODE REVIEW

**Comment by flcl42** [comment]
```
Environment.TickCount64 > DateTime.UtcNow > DateTime.Now
^ no sync/update issues > ^ no day-light saving time issues
```
I did not face such issues, but I believe if they will appear, it will be veery hard to debug

**Comment by LukaszRozmej** [comment]
> ```
> Environment.TickCount64 > DateTime.UtcNow > DateTime.Now
> ^ no sync/update issues   ^ no day-light saving time issues
> ```
> 
> I did not face such issues, but I believe if they will appear, it will be veery hard to debug

Yeah you are right. I somehow convinced myself that DateTime would not be a problem. Moved to DateTimeOffset.UtcNow as Environment.TickCount64 can have issues with resolution: https://docs.microsoft.com/en-us/dotnet/api/system.environment.tickcount?view=net-6.0#remarks

File: src/Nethermind/Nethermind.Merge.Plugin/BlockProduction/BlockImprovementContextFactory.cs [hunk 1]
```diff
@@ -32,10 +32,10 @@ public BlockImprovementContextFactory(IManualBlockProductionTrigger blockProduct
...
         _timeout = timeout;
     }
-    
-    public IBlockImprovementContext StartBlockImprovementContext(
-        Block currentBestBlock, 
-        BlockHeader parentHeader, 
-        PayloadAttributes payloadAttributes) =>
-        new BlockImprovementContext(currentBestBlock, _blockProductionTrigger, _timeout, parentHeader, payloadAttributes);
+
+    public IBlockImprovementContext StartBlockImprovementContext(Block currentBestBlock,
```
Comment (asdacap) [inline src/Nethermind/Nethermind.Merge.Plugin/BlockProduction/BlockImprovementContextFactory.cs:36]: NP: I'd prefer to keep this one the next line.

File: src/Nethermind/Nethermind.Merge.Plugin/BlockProduction/PayloadPreparationService.cs [hunk 2]
```diff
@@ -59,12 +71,17 @@ public class PayloadPreparationService : IPayloadPreparationService
...
+            int improvementDelay = DefaultImprovementDelayMs,
+            int minTimeForProduction = DefaultMinTimeForProductionMs)
         {
             _blockProducer = blockProducer;
             _blockImprovementContextFactory = blockImprovementContextFactory;
+            _timePerSlot = timePerSlot;
             TimeSpan timeout = timePerSlot;
-            _cleanupOldPayloadDelay = 2 * timePerSlot; // 2 * slots time
+            _cleanupOldPayloadDelay = 3 * timePerSlot; // 3 * slots time
+            _improvementDelay = TimeSpan.FromMilliseconds(improvementDelay);
```
Comment (asdacap) [inline src/Nethermind/Nethermind.Merge.Plugin/BlockProduction/PayloadPreparationService.cs:83]: NP: `improvementDelay` could have been a `TimeSpan`.

File: src/Nethermind/Nethermind.Merge.Plugin/BlockProduction/PayloadPreparationService.cs [hunk 3]
```diff
@@ -93,27 +110,70 @@ private Block ProduceEmptyBlock(string payloadId, BlockHeader parentHeader, Payl
...
 
-        private void ImproveBlock(string payloadId, BlockHeader parentHeader, PayloadAttributes payloadAttributes, Block emptyBlock)
+        private void ImproveBlock(string payloadId, BlockHeader parentHeader, PayloadAttributes payloadAttributes, Block currentBestBlock, DateTimeOffset startDateTime)
         {
-            if (_logger.IsTrace) _logger.Trace($"Start improving block from payload {payloadId} with parent {parentHeader}");
-            IBlockImprovementContext blockImprovementContext = _blockImprovementContextFactory.StartBlockImprovementContext(emptyBlock, parentHeader, payloadAttributes);
+            IBlockImprovementContext? oldContext = null;
+
+            _payloadStorage.AddOrUpdate(payloadId,
+                id => CreateBlockImprovementContext(id, parentHeader, payloadAttributes, currentBestBlock, startDateTime),
```
Comment (asdacap) [inline src/Nethermind/Nethermind.Merge.Plugin/BlockProduction/PayloadPreparationService.cs:118]: `oldContext` is not set if the context is new. Is that intended? If so, why not call `Dispose` in line 128 so that you can remove a var?

File: src/Nethermind/Nethermind.Merge.Plugin/BlockProduction/PayloadPreparationService.cs (hunk 2 above)
Comment (LukaszRozmej) [reply src/Nethermind/Nethermind.Merge.Plugin/BlockProduction/PayloadPreparationService.cs:83]: You mean constructor parameter? I was debating either keep it `int` in ms and have default value const, or make it `TimeSpan?` and have default value `null` and overwrite it in constructor with some `static readonly` field with proper default value. Thing is I want to keep those values optional in constructor. I decided on `int` to be more explicit, but don't have strong preference. Which do you think is better?

File: src/Nethermind/Nethermind.Merge.Plugin/BlockProduction/PayloadPreparationService.cs (hunk 3 above)
Comment (LukaszRozmej) [reply src/Nethermind/Nethermind.Merge.Plugin/BlockProduction/PayloadPreparationService.cs:115]: Right, I had to be focusing on other things.

File: src/Nethermind/Nethermind.Merge.Plugin/BlockProduction/PayloadPreparationService.cs (hunk 2 above)
Comment (asdacap) [reply src/Nethermind/Nethermind.Merge.Plugin/BlockProduction/PayloadPreparationService.cs:83]: I'd prefer if its a timespan, otherwise I'd have to read through to know that `improvementDelay` is in ms, or I have to rename it to `improvementDelayMs`. Not a big deal either way.

**Comment by LukaszRozmej** [comment]
MSF11 nodes looked, waiting for MSF12 good: https://seq.nethermind.io/#/events?filter=Contains(@Message,%20%22improved%22)%20and%20Logger%20%3D%20'Merge.Plugin.BlockProduction.PayloadPreparationService'

**Review by asdacap — APPROVED** [review]

**Review by marcindsobczak — APPROVED** [review]

**Comment by LukaszRozmej** [comment]
Sepolia tests showed no memory leak, everything is clearing up fine.

File: src/Nethermind/Nethermind.sln [hunk 4]
```diff
@@ -230,6 +230,8 @@ Project("{9A19103F-16F7-4668-BE54-9A1E7A4F7556}") = "Nethermind.Merge.AuRa.Test"
 EndProject
 Project("{9A19103F-16F7-4668-BE54-9A1E7A4F7556}") = "Nethermind.Merge.AuRa", "Nethermind.Merge.AuRa\Nethermind.Merge.AuRa.csproj", "{F166365D-6E31-401B-9753-59EADB800E2A}"
 EndProject
+Project("{FAE04EC0-301F-11D3-BF4B-00C04F79EFBC}") = "MathGmp.Native", "..\Math.Gmp.Native\MathGmp.Native\MathGmp.Native.csproj", "{16E039D0-8944-4EFB-8624-FC9EB9FF75DB}"
```
Comment (MarekM25) [inline src/Nethermind/Nethermind.sln:233]: Hmmm looks like not related to the PR

**Review by MarekM25 — APPROVED** [review]

File: src/Nethermind/Nethermind.sln (hunk 4 above)
Comment (LukaszRozmej) [reply src/Nethermind/Nethermind.sln:233]: yes but fixes latest master
